hasktorch/src/Torch/TensorFactories.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module Torch.TensorFactories where
+
+import System.IO.Unsafe
+import Foreign.ForeignPtr
+
+import qualified ATen.Const as ATen
+import qualified ATen.Managed.Native as ATen
+import qualified ATen.Managed.Type.TensorOptions as ATen
+import qualified ATen.Type as ATen
+import qualified Torch.Managed.Native as LibTorch
+import ATen.Managed.Cast
+import ATen.Class (Castable(..))
+import ATen.Cast
+
+import Torch.Tensor
+import Torch.TensorOptions
+
+-- XXX: We use the torch:: constructors, not at:: constructures, because
+--      otherwise we cannot use libtorch's AD.
+
+type FactoryType = ForeignPtr ATen.IntArray
+                    -> ForeignPtr ATen.TensorOptions
+                    -> IO (ForeignPtr ATen.Tensor)
+
+mkFactory :: FactoryType -> [Int] -> TensorOptions -> IO Tensor
+mkFactory aten_impl shape opts = (cast2 aten_impl) shape opts
+
+mkFactoryUnsafe :: FactoryType -> [Int] -> TensorOptions -> Tensor
+mkFactoryUnsafe f shape opts = unsafePerformIO $ mkFactory f shape opts
+
+mkDefaultFactory :: ([Int] -> TensorOptions -> a) -> [Int] -> a
+mkDefaultFactory non_default shape = non_default shape defaultOpts
+
+-------------------- Factories --------------------
+
+ones :: [Int] -> TensorOptions -> Tensor
+ones = mkFactoryUnsafe LibTorch.ones_lo
```

Comment:
I think not using unsafePerformIO is bertter for factory method.
The behavior changes by optimization. See a following example.
By optimization, the number of calling unsafePerformIO is reduced.

```
$ cat test.hs
import System.IO.Unsafe
ones :: Int -> Int
ones p = unsafePerformIO $ do
  print p
  return p
main = do
  let a = ones 1
      b = ones 1
  print a
  print b
$ ghc test.hs
$ ./test
1
1
1
1
$ ghc test.hs -O2
$ ./test
1
1
1
```

Reply:
Good point, I'll update that.

Reply:
I have written the side-effect of unsafePerformIO  for hasktorch.
https://github.com/junjihashimoto/ffi-experimental/blob/hasktorch-with-test/hasktorch/test/GradSpec.hs
The code is below.

```
spec :: Spec
spec = do
  it "grad with ones" $ do
    let x = ones [] $ withRequiresGrad defaultOpts
        y = x * x + 5 * x + 3
    fmap asDouble (grad y [x]) `shouldBe` [7.0]
  it "grad with ones" $ do
    let x = ones [] $ withRequiresGrad defaultOpts
        y = ones [] $ withRequiresGrad defaultOpts
        z = x * x * y
    fmap asDouble (grad z [x]) `shouldBe` [2.0]
    fmap asDouble (grad z [y]) `shouldBe` [1.0]
```

The test is blow.
By optimization, y becomes x.
We expect 2, but the result is 3, because grad (x*x*x) [x]==3.

```
hasktorch-0.2.0.0: test (suite: spec)


Grad
  grad with ones
  grad with ones FAILED [1]

Failures:

  test/GradSpec.hs:23:5:
  1) Grad grad with ones
       expected: [2.0]
        but got: [3.0]
```